master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala

```diff
@@ -653,11 +676,36 @@ private[celeborn] class Master(
       override def run(): Unit = {
         statusSystem.handleAppLost(appId, requestId)
         logInfo(s"Removed application $appId")
+        if (conf.hasHDFSStorage) {
+          cleanExpiredAppDirsOnHDFS(appId)
+        }
         context.reply(ApplicationLostResponse(StatusCode.SUCCESS))
       }
     })
   }
 
+  private def cleanExpiredAppDirsOnHDFS(dir: String = ""): Unit = {
+    if (hadoopFs == null) {
+      hadoopFs = CelebornHadoopUtils.getHadoopFS(conf)
+    }
+    val hdfsWorkPath = new Path(conf.hdfsDir, conf.workerWorkingDir)
+    if (hadoopFs.exists(hdfsWorkPath)) {
+      if (!dir.isEmpty) {
+        // delete specific app dir on application lost
+        hadoopFs.delete(new Path(hdfsWorkPath, dir), true)
+      } else {
+        val iter = hadoopFs.listStatusIterator(hdfsWorkPath)
+        while (iter.hasNext) {
+          val fileStatus = iter.next()
+          if (!statusSystem.appHeartbeatTime.containsKey(fileStatus.getPath.getName)) {
+            logDebug(s"Clean HDFS dir ${fileStatus.getPath.toString}")
```

Comment:
info, and maybe we should log time cost too, sometimes it's time-consuming

Reply:
OK, changed.